cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -219,48 +219,89 @@ struct MultiplyChecked {
   }
 };
 
+struct Divide {
+  template <typename T>
+  static constexpr enable_if_floating_point<T> Call(KernelContext*, T left, T right) {
+    return left / right;
+  }
+
+  template <typename T>
+  static constexpr enable_if_integer<T> Call(KernelContext*, T left, T right) {
+    return left / right;
+  }
+};
+
+struct DivideChecked {
+  template <typename T>
+  static enable_if_integer<T> Call(KernelContext* ctx, T left, T right) {
+    if (right == 0) {
+      ctx->SetStatus(Status::Invalid("divide by 0"));
+      return right;
+    }
+    return Divide::Call(ctx, left, right);
+  }
+
+  template <typename T>
+  static constexpr enable_if_floating_point<T> Call(KernelContext*, T left, T right) {
+    return left * right;
+  }
+};
+
 using applicator::ScalarBinaryEqualTypes;
+using applicator::ScalarBinaryNotNullEqualTypes;
 
 // Generate a kernel given an arithmetic functor
 //
 // To avoid undefined behaviour of signed integer overflow treat the signed
 // input argument values as unsigned then cast them to signed making them wrap
 // around.
 template <typename Op>
-ArrayKernelExec NumericEqualTypesBinary(detail::GetTypeId get_id) {
+ArrayKernelExec NumericEqualTypesBinary(detail::GetTypeId get_id,
+                                        bool skip_null = false) {
   switch (get_id.id) {
     case Type::INT8:
-      return ScalarBinaryEqualTypes<Int8Type, Int8Type, Op>::Exec;
+      return skip_null ? ScalarBinaryNotNullEqualTypes<Int8Type, Int8Type, Op>::Exec
```

Comment:
I think this will cause both templates to always be instantiated, which is not what we want